feat: grouped asset by primary asset

NeOMakinG opened the pull request:
## Description

<!-- Please describe your changes -->

## Issue (if applicable)

<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

closes #

## Risk

> High Risk PRs Require 2 approvals

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

WARNING: If your PR introduces a new on-chain transaction or modifies an existing one, please add the "High-Risk" label to this PR and ask for 2 reviewers to approve it before merging.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing

<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>

### Engineering

<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

## Screenshots (if applicable)

premiumjibles commented on src/components/AssetSearch/components/GroupedAssetRow.tsx:
```diff
@@ -0,0 +1,273 @@
+        </Box>
+      )
+    })
+  }, [relatedAssets])
```
Agree with code rabbit here. Wherever it makes sense I think it's safer to avoid doing a store.getState(). We might accidentally use it somewhere where we need reactivity

NeOMakinG replied:
Done

premiumjibles commented on src/components/StakingVaults/Cells.tsx:
```diff
@@ -65,6 +71,10 @@ const buildRowTitle = (asset: Asset, postFix?: string, showAssetSymbol?: boolean
   }
 
+  if (isGrouped) {
+    return asset.name.split(' on ')[0] ?? asset.name
```
Don't love this but agree it's probably the best we've got without major changes. Potentially we could run a common prefix function over the related asset names? Not sure if that's going to be more or less reliable

NeOMakinG replied:
Done in another PR

premiumjibles commented on src/pages/Dashboard/components/AccountList/GroupedAssetsSubComponent.tsx:
```diff
@@ -0,0 +1,111 @@
+    vibrate('heavy')
+    onRowLongPress?.(row as Row<AccountRowData>)
+  }, defaultLongPressConfig)
+
```
Might be what's breaking long press

NeOMakinG replied:
I don't think it was 100% of the issue, but made longpress great again!

premiumjibles reviewed (changes requested):
Overall looks to work well in all the places it's implemented 👌 . Much more intuitive.


Blocking: Long press to get the context menu on mobile view doesn't seem to be working any more (at least on web view mobile).

Maybe blocking: Are we supposed to have the grouping on the assets route?

<img width="527" height="671" alt="Image" src="https://github.com/user-attachments/assets/9bfde554-f38e-47c2-a934-98bc542c7d9c" />

For some reason when I do a global search i'm getting different results between prd and this branch. Confirmed if i'm on local develop it matches prod, so something related to how this branch is doing search results.

<img width="731" height="388" alt="Image" src="https://github.com/user-attachments/assets/b6461bdb-44a2-4761-987f-f2969bdcecdb" />

<img width="601" height="726" alt="Image" src="https://github.com/user-attachments/assets/26a680c5-5419-4694-8bd6-5444f6c9c9f4" />


I'd be inclined to bump this one up to high risk and get a second review. Just by nature the surface area here is massive (which also means high chance of a million different edge cases) and is touching some important stuff. 

Code wise there are some code rabbit suggestions that look reasonable or worth looking into but I don't have many specific nitpicks. Looks sane.

From a product perspective I do wonder if rendering all these icons is a little bit hectic but I don't feel strongly. I am slightly concerned on rendering performance but we'll see how it goes in a built env. 

<img width="472" height="932" alt="Image" src="https://github.com/user-attachments/assets/f9bd8fe3-3089-4f2d-b625-2f98d1fbe97e" />

NeOMakinG commented on src/components/ReactTable/InfiniteTable.tsx:
```diff
@@ -96,141 +80,220 @@ export const InfiniteTable = <T extends {}>({
+        </Th>
+      )),
+    )
+  }, [displayHeaders, headerGroups, hoverColor, translate])
```
We don't have a TR there why would we take car of the key if virtuoso is already? Even more nitpick as it's headers

NeOMakinG commented on src/pages/Dashboard/components/AccountList/AccountTable.tsx:
```diff
@@ -43,18 +53,56 @@ const emptyContainerProps: FlexProps = {
   const sortedRows = useMemo(() => {
     return rowData.sort((a, b) => Number(b.fiatAmount) - Number(a.fiatAmount))
   }, [rowData])
   const { hasMore, next, data } = useInfiniteScroll({
```
Why not, even thought in the future can you review actual changes only as I didn't change this part?

NeOMakinG commented:
> Overall looks to work well in all the places it's implemented 👌 . Much more intuitive.
> 
> Blocking: Long press to get the context menu on mobile view doesn't seem to be working any more (at least on web view mobile).
> 
> Maybe blocking: Are we supposed to have the grouping on the assets route?
> 
> <img alt="Image" width="527" height="671" src="https://private-user-images.githubusercontent.com/16397089/484343607-9bfde554-f38e-47c2-a934-98bc542c7d9c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY4MTY2ODYsIm5iZiI6MTc1NjgxNjM4NiwicGF0aCI6Ii8xNjM5NzA4OS80ODQzNDM2MDctOWJmZGU1NTQtZjM4ZS00N2MyLWE5MzQtOThiYzU0MmM3ZDljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTAyVDEyMzMwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI0MWMxZTI1M2MzMzlmNjBhNmRhZWQ2NDRiZjgzYjk5MzU2NzFhYjQ1ODVhODgyZmZiNGNhYjdjYzg1ODIwYzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.BzO9kKO6XGJmb0ELhIWj1zlIzG1ZfvxdCiaSuHSrjU4">
> For some reason when I do a global search i'm getting different results between prd and this branch. Confirmed if i'm on local develop it matches prod, so something related to how this branch is doing search results.
> 
> <img alt="Image" width="731" height="388" src="https://private-user-images.githubusercontent.com/16397089/484350129-b6461bdb-44a2-4761-987f-f2969bdcecdb.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY4MTY2ODYsIm5iZiI6MTc1NjgxNjM4NiwicGF0aCI6Ii8xNjM5NzA4OS80ODQzNTAxMjktYjY0NjFiZGItNDRhMi00NzYxLTk4N2YtZjI5NjliZGNlY2RiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTAyVDEyMzMwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE2OWFhMTE0NjgwOGE1MzI4NTYzY2IyMTViMTY3OTU3OTQ3MGFmNTA0MGEyYTE5YjFiNmJkYzIxMTdmYjc0NjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.UfKHR9v8Q-8qFLiYgNjo2FDxQJM6US7EtO8Ar0gex5g"> <img alt="Image" width="601" height="726" src="https://private-user-images.githubusercontent.com/16397089/484350146-26a680c5-5419-4694-8bd6-5444f6c9c9f4.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY4MTY2ODYsIm5iZiI6MTc1NjgxNjM4NiwicGF0aCI6Ii8xNjM5NzA4OS80ODQzNTAxNDYtMjZhNjgwYzUtNTQxOS00Njk0LThiZDYtNTQ0NGY2YzljOWY0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTAyVDEyMzMwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVlZjRhMTUyM2RjYTdjM2EyYmU2NmRhNmFjY2Y5YjY3NGIxNTg0NWM4MGViYjg0MzZmYWE0ODk3NGM2YjQwNzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cTKFgXiLe-Lo7Y8waa1DibdJrqFskt86-jR6SgmVFX4">
> I'd be inclined to bump this one up to high risk and get a second review. Just by nature the surface area here is massive (which also means high chance of a million different edge cases) and is touching some important stuff.
> 
> Code wise there are some code rabbit suggestions that look reasonable or worth looking into but I don't have many specific nitpicks. Looks sane.
> 
> From a product perspective I do wonder if rendering all these icons is a little bit hectic but I don't feel strongly. I am slightly concerned on rendering performance but we'll see how it goes in a built env.
> 
> <img alt="Image" width="472" height="932" src="https://private-user-images.githubusercontent.com/16397089/484350696-f9bd8fe3-3089-4f2d-b625-2f98d1fbe97e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY4MTY2ODYsIm5iZiI6MTc1NjgxNjM4NiwicGF0aCI6Ii8xNjM5NzA4OS80ODQzNTA2OTYtZjliZDhmZTMtMzA4OS00ZjJkLWI2MjUtMmY5OGQxZmJlOTdlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTAyVDEyMzMwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM5MmM5ZTZhZmMwZGNkN2Y4MzBkOGU3NWRjOGJiNjA0N2E1ZDBkZmMwOGYwYTE0MDZhMWZjNzc3MDNiNjRmNTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.17sYmRkbBahk5gjIYtq286VXcAwDt8QlNrznOt2Iud4">

Thanks for the catches, it should be addressed now

I don't know why the research was borken, but it's working fine now, probably because of sorting things we had, I removed an extra step which was not needed and took care of other issues, fixed sorting...

Changed to high and I'll invoke gomes for a big review as I know his eyes will catch a lot of weird things for sure

NeOMakinG commented on src/components/AssetSearch/components/AssetList.tsx:
```diff
@@ -42,47 +49,71 @@ export const AssetList: FC<AssetListProps> = ({
+    const grouped = groupBy(assets, 'relatedAssetKey')
+    return Object.values(grouped).map(group => group[0])
+  }, [assets, shouldDisplayRelatedAssets])
+
```
I'm not using grouping anymore as it was breaking the sorting

NeOMakinG commented on src/components/AssetSearch/components/GroupedAssetRow.tsx:
```diff
@@ -0,0 +1,273 @@
+                  value={totalBalanceBaseUnit}
+                  symbol={primaryAsset.symbol}
+                />
+              ) : (
```
Done

NeOMakinG commented on src/components/AssetSearch/components/GroupedAssetRow.tsx:
```diff
@@ -0,0 +1,273 @@
+                color='var(--chakra-colors-chakra-body-text)'
+                value={totalBalance.toString()}
+              />
+              <Flex>{networksIcons}</Flex>
```
Done

NeOMakinG commented on src/components/Table/Table.theme.ts:
```diff
@@ -168,6 +160,10 @@ export const TableStyle = {
         tbody: {
           tr: {
             borderRadius: 'xl',
+            '.expanded': {
+              borderBottomRightRadius: '0',
+              borderBottomLeftRadius: '0',
+            },
             _focus: {
```
I think it works, prefer not to take more time here

NeOMakinG commented on src/components/ReactTable/InfiniteTable.tsx:
```diff
@@ -96,141 +82,230 @@ export const InfiniteTable = <T extends {}>({
+      onVisibleRowsChange?.(visibleRows.map(row => row.original))
+    },
+    [preparedRows, onVisibleRowsChange],
+  )
```
Fixed

NeOMakinG commented on src/components/MarketTableVirtualized/MarketsTableVirtualized.tsx:
```diff
@@ -83,31 +39,36 @@ export const MarketsTableVirtualized: React.FC<MarketsTableVirtualizedProps> = m
+        const baseAssetName = asset.name.split(' on ')[0]
+        return `${baseAssetName}_${asset.symbol}`
+      })
+
```
It's working prefer not to touch anymore considering the big PR already, waiting for next review

gomesalexandre commented on src/components/AssetSearch/components/GroupedAssetRow.tsx:
```diff
@@ -0,0 +1,301 @@
+      selectPortfolioCryptoPrecisionBalanceByFilter(state, { assetId: id }),
+    ),
+  )
+  const totalBalanceCryptoHuman = useMemo(
```
q: couldn't we do this only once with crypto and prorate with market data user currency? That would keep things a lil bit simpler (assuming all variants have the same price over all chains)

gomesalexandre commented on src/components/AssetSearch/components/GroupedAssetRow.tsx:
```diff
@@ -0,0 +1,301 @@
+
+  const changePercent24Hr = marketData?.changePercent24Hr
+
+  const changePercentTagColorsScheme = useMemo(() => {
```
ditto extract?

gomesalexandre commented on src/components/MarketTableVirtualized/MarketsTableVirtualized.tsx:
```diff
@@ -16,65 +15,22 @@ import { SparkLineCell } from './SparkLineCell'
+  onRowLongPress?: (asset: Asset) => void
+}
+
+type AssetWithRelatedAssetIds = Asset & {
+  isGrouped?: boolean
+  relatedAssetIds?: string[]
```
- Shouldn't `relatedAssetIds` live alongside `Asset` itself? 
- `isGrouped` makes me think this should probably be named something along the lines of asset group

gomesalexandre commented on src/components/MarketTableVirtualized/MarketsTableVirtualized.tsx:
```diff
@@ -124,36 +85,40 @@ export const MarketsTableVirtualized: React.FC<MarketsTableVirtualizedProps> = m
+            <AssetCell
+              assetId={row.original.assetId}
+              symbol={row.original.symbol}
+              isGrouped={Boolean(row.original.isGrouped)}
```
```suggestion
              isGrouped={row.original.isGrouped}
```

since `<AssetCell />` is happy with undefined, can do the falsy dancy

NeOMakinG replied:
Not needed as it will be managed by another PR, removed it

gomesalexandre commented on src/pages/Explore/components/CategoryCard.tsx:
```diff
@@ -171,22 +172,40 @@ export const CategoryCard = memo(
+              showPrice
             />
-          ))}
+          ))}{' '}
```
eslint rug?

NeOMakinG replied:
Fixed

gomesalexandre commented on src/components/ReactTable/InfiniteTable.tsx:
```diff
@@ -96,141 +82,231 @@ export const InfiniteTable = <T extends {}>({
+
+  const handleRowToggle = useCallback((rowId: string) => {
+    setExpandedItems(prev => {
+      const newSet = new Set(prev)
```
can we pls give this akschual names 🙏🏽

gomesalexandre commented on src/pages/Dashboard/components/AccountList/AccountTable.tsx:
```diff
@@ -43,18 +53,52 @@ const emptyContainerProps: FlexProps = {
+    })
+
+    const groupedResults = Object.values(grouped).map(group => {
+      if (group.length === 1) return group[0]
```
q: this applies for all occurrences of `group[0]` in this PR but am I correct in assuming we *could* run into the wrong primary asset, but that's actually not an issue, since as far as this PR is concerned, a primary asset is just used to display one of many implementations, and we're not so concerned about it being the "correct" one (e.g ETH on Ethereum) but rather it being the row?

NeOMakinG replied:
Not existing anymore, we now rely on the primary related asset

NeOMakinG commented on src/lib/utils/asset.ts:
```diff
@@ -0,0 +1,5 @@
+import type { AssetId } from '@shapeshiftoss/caip'
+
+export const isPrimaryAsset = (relatedAssetKey: AssetId | null | undefined, assetId: AssetId) => {
+  return relatedAssetKey === null || relatedAssetKey === assetId
+}
```
No, undefined means the asset doesn't have related assets, it's considered as an asset but not a primary asset, primary asset notion is when asset as some underlying related assets

NeOMakinG commented on src/pages/Dashboard/components/AccountList/AccountTable.tsx:
```diff
@@ -65,6 +76,10 @@ export const AccountTable = memo(() => {
           <AssetCell
             assetId={row.original.assetId}
             subText={truncate(row.original.symbol, { length: 6 })}
+            isGrouped={Boolean(
+              !row.original.relatedAssetKey ||
+                row.original.relatedAssetKey === row.original.assetId,
+            )}
           />
```
I'll remove it in the next commit

NeOMakinG commented on src/state/slices/assetsSlice/selectors.ts:
```diff
@@ -41,6 +42,14 @@ export const selectAssets = createDeepEqualOutputSelector(
   byId => byId,
 )
 
+export const selectPrimaryAssets = createCachedSelector(assets.selectors.selectAssetsById, byId => {
+  return Object.values(byId).reduce<AssetsById>((acc, asset) => {
+    if (asset && isPrimaryAsset(asset.relatedAssetKey, asset.assetId)) acc[asset.assetId] = asset
+
+    return acc
+  }, {})
+})(() => 'primaryAssets')
```
It's supposed to be null only looking at asset generation

NeOMakinG commented on src/components/AssetSearch/components/AssetRow.tsx:
```diff
@@ -44,10 +77,135 @@ export const AssetRow: FC<{ asset: Asset; index: number; data: AssetData }> = me
+          />
+        </Tag>
+      )
+    }, [changePercent24Hr, changePercentTagColorsScheme])
```
Added a component and reused everywhere

NeOMakinG commented on src/components/ReactTable/InfiniteTable.tsx:
```diff
@@ -96,141 +82,236 @@ export const InfiniteTable = <T extends {}>({
+  )
+
+  if (isLoading) {
+    return (
+      <Table variant={variant} size={tableSize} className='infinite-table'>
+        {renderHeader()}
+        <Tbody>
+          {Array.from({ length: 5 }).map((_, index) => (
+            <Tr key={index}>
+              {columns.map((_column, colIndex) => (
+                <Td key={colIndex}>
+                  <Skeleton height='16px' />
+                </Td>
+              ))}
+            </Tr>
+          ))}
+        </Tbody>
       </Table>
-    </InfiniteScroll>
+    )
+  }
```
nop it has a thead already

NeOMakinG commented on src/components/PriceChangeTag/PriceChangeTag.tsx:
```diff
@@ -0,0 +1,38 @@
+    return 'gray'
+  }, [changePercent24Hr])
+
+  if (!changePercent24Hr) return null
+
+  return (
+    <Tag colorScheme={changePercentTagColorsScheme} width='max-content' px={1} size='sm'>
+      {changePercentTagColorsScheme !== 'gray' ? (
+        <TagLeftIcon
+          as={changePercentTagColorsScheme === 'green' ? RiArrowRightUpLine : RiArrowLeftDownLine}
+          me={1}
+        />
+      ) : null}
+      <Amount.Percent value={bnOrZero(changePercent24Hr).times('0.01').toString()} fontSize='xs' />
+    </Tag>
```
I want to hide 0% changes

NeOMakinG commented on src/components/TradeAssetSearch/hooks/useGetPopularAssetsQuery.tsx:
```diff
@@ -17,17 +18,32 @@ export const queryFn = async () => {
   if (enabledFlags.Mayachain) assetIds.push(mayachainAssetId)
 
   for (const assetId of assetIds) {
-    const asset = assets[assetId]
+    const asset = primaryAssets[assetId]
+    const relatedAssetIds = selectRelatedAssetIds(store.getState(), { assetId })
     if (!asset) continue
     const { chainId } = asset
     if (!result[chainId]) result[chainId] = []
```
We want primary assets, if a sub asset is in the related asset ids we will include the primary asset related to it don't worry

NeOMakinG commented on src/components/TradeAssetSearch/workers/assetSearch.worker.ts:
```diff
@@ -3,6 +3,8 @@
   Receives an initial dataset of minimal assets, caches it,
   and responds to search requests with filtered assetIds.
 */
+import type { AssetId } from '@shapeshiftoss/caip/dist/cjs'
```
thx will be fixed in next commit

NeOMakinG commented:
thanks for the big review @gomesalexandre as always

Should be good now hopefully, I fixed everything noticed minus the flashing of row opening on account/assets page as I took ilke 3-4 hours trying to find a fix and couldn't find one, I would love to give this a try as a follow up later

gomesalexandre commented on src/state/slices/portfolioSlice/selectors.ts:
```diff
@@ -723,6 +726,8 @@ export const selectAccountIdsByAssetIdAboveBalanceThresholdByFilter = createDeep
       : accountIdsAboveThreshold,
 )
 
+export type RowProps = Row<AccountRowData>
```
Can we name this `AccountRowProps` and give it a better, less generic home?

NeOMakinG replied:
Done in [09cab7e](https://github.com/shapeshift/web/pull/10380/commits/09cab7eda0adfea64fc9a6d343482264710f73cc)

gomesalexandre commented on src/components/PriceChangeTag/PriceChangeTag.tsx:
eyy why not

gomesalexandre commented on packages/types/src/base.ts:
```diff
@@ -85,12 +85,13 @@ export type Asset = {
   explorerAddressLink: string
   relatedAssetKey: AssetId | null | undefined
+  isPrimary?: boolean
+  isChainSpecific?: boolean
```
q: wat dis?
